osu.Game/Online/RealtimeMultiplayer/StatefulMultiplayerClient.cs

```diff
@@ -0,0 +1,389 @@
+        }
+
+        Task IMultiplayerClient.MatchStarted()
+        {
+            Debug.Assert(Room != null);
+            var players = Room.Users.Where(u => u.State == MultiplayerUserState.Playing).Select(u => u.UserID).ToList();
+
+            Schedule(() =>
+            {
+                if (Room == null)
+                    return;
+
+                PlayingUsers.AddRange(players);
+
+                MatchStarted?.Invoke();
+            });
+
+            return Task.CompletedTask;
+        }
+
+        Task IMultiplayerClient.ResultsReady()
+        {
+            Schedule(() =>
+            {
+                if (Room == null)
+                    return;
+
+                ResultsReady?.Invoke();
+            });
+
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Populates the <see cref="User"/> for a given <see cref="MultiplayerRoomUser"/>.
+        /// </summary>
+        /// <param name="multiplayerUser">The <see cref="MultiplayerRoomUser"/> to populate.</param>
+        protected async Task PopulateUser(MultiplayerRoomUser multiplayerUser) => multiplayerUser.User ??= await userLookupCache.GetUserAsync(multiplayerUser.UserID);
+
+        /// <summary>
+        /// Updates the local room settings with the given <see cref="MultiplayerRoomSettings"/>.
+        /// </summary>
+        /// <remarks>
+        /// This updates both the joined <see cref="MultiplayerRoom"/> and the respective API <see cref="Room"/>.
+        /// </remarks>
+        /// <param name="settings">The new <see cref="MultiplayerRoomSettings"/> to update from.</param>
+        private void updateLocalRoomSettings(MultiplayerRoomSettings settings)
+        {
+            if (Room == null)
+                return;
+
+            // Update a few instantaneously properties of the room.
+            Schedule(() =>
+            {
+                if (Room == null)
+                    return;
+
+                Debug.Assert(apiRoom != null);
+
+                Room.Settings = settings;
+                apiRoom.Name.Value = Room.Settings.Name;
+
+                // The playlist update is delayed until an online beatmap lookup (below) succeeds.
+                // In-order for the client to not display an outdated beatmap, the playlist is forcefully cleared here.
+                apiRoom.Playlist.Clear();
+
+                RoomChanged?.Invoke();
+            });
+
+            var req = new GetBeatmapSetRequest(settings.BeatmapID, BeatmapSetLookupType.BeatmapId);
+            req.Success += res =>
+            {
+                var beatmapSet = res.ToBeatmapSet(rulesets);
+
+                var beatmap = beatmapSet.Beatmaps.Single(b => b.OnlineBeatmapID == settings.BeatmapID);
+                beatmap.MD5Hash = settings.BeatmapChecksum;
+
+                var ruleset = rulesets.GetRuleset(settings.RulesetID);
+                var mods = settings.Mods.Select(m => m.ToMod(ruleset.CreateInstance()));
```

Comment:
The ruleset instance should probably created once here instead of once per mod. There's an underlying reflection call, so not 100% negligible I'd say.